woocommerce/woocommerce-android
# Shipping Label Packages: Step 1

Implements the first task and some parts of the second task of #3019, which is displaying the list of products when trying to edit packages of a shipping label, and displays a spinner for editing the selected package.

I felt a bit lost when working on this, and maybe there are some easier ways or better solutions than what I used, so please tell me about them 🙂.

I'll try to explain some decisions that I made here:

1. Loading packages is done in `CreateShippingLabelsViewModel` and not `EditShippingLabelPackagesViewModel`, I did this because I want to cache those packages, to avoid reloading them if the user wants to edit their changes later. I couldn't cache them in the `ShippingLabelRepository`, as I would have to make it a singleton, but this means we will cache them for the app lifecycle.
2. I combined both the custom packages, and predefined one in the same app model: `ShippingPackage`, this makes handling the list easier, and we can differentiate between them using the `category` field.
3. `ShippingPackage` has an `id` field, this is not used now, but we will need it to fetch the rates, and it has to be added in FluxC, I'll do it in the upcoming PRs.
4. The `EditShippingLabelPackagesFragment` may show a skeleton to fetch the order products if they weren't cached before, and in case of errors during fetching, we will just ignore the error and hide the weight, this was confirmed by Garance.

<img src="https://user-images.githubusercontent.com/1657201/106260812-1cb50400-6221-11eb-9a43-388653555279.png" width=400/>

Both Unit tests, and the rest of the task will be done in upcoming PRs, as this is already too big.

#### Testing

- Prepare an order on the web that satisfies SL creation preconditions.
- Go to Orders -> Open the order
- Pass the steps of origin and destination addresses.
- Click on Continue of Package details.
- Notice that a loading dialog appears, and then the package details screen will be displayed.
- Go Back.
- Click on Continue a second time.
- Notice the package details screen is opened without reloading packages.

Update release notes:
